source/shared/smart-contracts/tutorials/piggy-bank/writing.rst

```diff
@@ -429,65 +430,88 @@ an account address that is equal to the owner:
 
 .. code-block:: rust
 
-   ensure!(*state == PiggyBankState::Intact);
+   ensure!(*host.state() == PiggyBankState::Intact);
 
 At this point you know the piggy bank is still intact and the sender is the
-owner, meaning you now get to the smashing part:
+owner, meaning you now get to the smashing part.
+But there is one problem.
+The state is immutable, so we first need to make the receive function mutable by
+adding the |mutable|_ attribute to the |receive|_ macro.
 
 .. code-block:: rust
+   :emphasize-lines: 1, 4
+
+   #[receive(contract = "PiggyBank", name = "smash", mutable)]
+   fn piggy_smash<S: HasStateApi>(
+       ctx: &impl HasReceiveContext,
+       host: &mut impl HasHost<PiggyBankState, StateApiType = S>,
+   ) -> ReceiveResult<()> {
+       let owner = ctx.owner();
+       let sender = ctx.sender();
+       ensure!(sender.matches_account(&owner));
+       ensure!(*host.state() == PiggyBankState::Intact);
+
+       todo!()
+   }
 
-   *state = PiggyBankState::Smashed
+This gives us a mutable reference to the ``host``, through which we can access
+the mutable state with the ``state_mut`` function. We then set the state to
+``Smashed``, preventing further insertions of CCD:
 
-Since the state is a mutable reference, you can simply mutate it to be
-``Smashed``, preventing anyone from inserting any more CCD.
+.. code-block:: rust
+
+   *host.state_mut() = PiggyBankState::Smashed;
 
-Lastly you need to empty the piggy bank. To do that, transfer all the CCD
+Lastly, you need to empty the piggy bank. To do that, transfer all the CCD
 of the smart-contract instance to an account.
 
-To transfer CCD from a smart contract instance you create an
-action for a simple transfer, again using the generic ``A``.
-To construct a simple transfer you need to provide the address of the receiving
-account and the amount to be transferred.
+To transfer CCD from a smart contract instance you use the |invoke_transfer|_
+method on the ``host``. For this, you need to provide the address of the receiving
+account and the amount to transfer.
 In this case the receiver is the owner of the piggy bank and the amount is the
 entire balance of the piggy bank.
 
-The context has a getter function for reading
+The ``host`` has a getter function for reading
 the current balance of the smart contract instance, which is called
 |self_balance|_:
 
 .. code-block:: rust
 
-   let balance = ctx.self_balance();
+   let balance = host.self_balance();
 
 You already have a variable with the address of the contract owner, so you can
-construct and return the action for a simple transfer:
+use that for invoking the transfer:
 
 .. code-block:: rust
 
-   Ok(A::simple_transfer(&owner, balance))
+   Ok(host.invoke_transfer(&owner, balance)?)
+
+A transfer can fail in two ways, either your contract has insufficient funds, or
+the receiver account does not exist. Neither can occur in this contract, so we
```

Comment:
Perhaps explain why neither can occur.

A natural question a reader might ask is why propagate, instead of panic.